### ISSUE TITLE
Add op-assign instructions.

### ISSUE THREAD

**Issue by Robbepop** [paritytech/wasmi#767]
Follow-up from https://github.com/paritytech/wasmi/pull/729.

Currently the `wasmi` register-machine bytecode only consists of normal 3-way instructions in the form: `op(result, input0, input1)` where `result` and `input0` or `input1` can overlap. An example is `i32.add $result $lhs $rhs`.

The idea in this issue and optimization is to introduce op-assign instructions such as `i32.add_assign $inout $input`.

The advantage is that we no longer have to store both, `result` and two input fields since `result` and `lhs` are always the same and always of type `Register`. This also allows to have inline 32-bit encoded constant values in the `rhs` field and thus we could save some encoding space, too.
Another benefit is that the instruction is probably also a bit faster during execution since the compiler has more information about which registers to read to and write from. However, this has to be checked with experiments.

A downside of having these op-assign instructions is that they do not play well with `local.set` and `local.tee` optimizations where we replace the result register of the previous instruction during translation phase.

Examples of new instructions are:

```rust
enum Instruction {
    ...
    I32AddAssign {
        inout: Register,
        rhs: Register,
    },
    I32AddAssignImm {
        inout: Register,
        rhs: Const32<i32>,
    },
    ...
}
```

Another important factor concerning optimization is that with these new instructions there is a chance that modern branch predictors can do an even better job when being able to distinguishing between `i32.add` and `i32.add_assign` instructions during execution to predict the next instruction in the sequence.

**Comment by Robbepop** [comment]
Today I experimentally added `Instruction::I32AddAssignImm` to the register-machine `wasmi` to see how much this single new op-assign instruction would improve performance of our existing benchmarks.

I tested this specifically for our `count_until` and `tiny_keccak` benchmark tests.

- `count_until`: Improved only marginally and within error. I guess the branch instructions dominate the execution time.
- `tiny_keccak`: Improved by roughly 2-4%. This might indicate that there is a lot to gain if this optimization is applied across the board.

The branch where I conducted the experiment can be found here: https://github.com/paritytech/wasmi/tree/rf-op-assign-instrs
